File: TDD/TDD Coding Practice Arithmetic/Double Stack Algorithm/Code/Java/src/fayelab/tdd/arithmetic/doublestack/original/Expression.java
package fayelab.tdd.arithmetic.doublestack.original;

public class Expression
{
    private Context ctx;
    private DoubleStackProcessor dsp;

    public Expression(String expr)
    {
        this(new Context(expr, 0));
    }

    Expression(Context ctx)
    {
        this.ctx = ctx;
        this.dsp = new DoubleStackProcessor();
    }

    public int eval()
    {
        while(ctx.exprNotEnd())
        {
            char c = ctx.nextChar();
            
            if('(' == c)
            {
                dsp.pushOperand(new Expression(ctx).eval());
            }
            else if(')' == c)
            {
                return dsp.result();
            }
            else
            {
                dsp.process(c);
            }
        }
        
        return dsp.result();
    }
}

class Context
{
    private String expr;
    private int curIdx;
    
    Context(String expr, int curIdx)
    {
        this.expr = expr;
        this.curIdx = curIdx;
    }
    
    boolean exprNotEnd()
    {
        return curIdx < expr.length();
    }

    char nextChar()
    {
        return expr.charAt(curIdx++);
    }
}

File: TDD/TDD Coding Practice Arithmetic/Double Stack Algorithm/Code/Python/oo/original/expression.py
import doublestackprocessor as dsp

def eval_expr(expr):
    return _eval_expr(expr, 0, [], [])

def _eval_expr(expr, curidx, operandstack, operatorstack):
    while not_end(expr, curidx):
        c, curidx = next_char(expr, curidx)
        if '(' == c:
            dsp.push_operand(_eval_expr(expr, curidx, [], []))
        elif ')' == c:
            return dsp.result(operandstack, operatorstack)
        else:
            dsp.process(c, operandstack, operatorstack)

    return dsp.result(operandstack, operatorstack)

def not_end(expr, curidx):
    return curidx < len(expr)

def next_char(expr, curidx):
    c = expr[curidx]
    curidx += 1
    return c, curidx
